### ISSUE TITLE
Envoy proxy is NOT ready: config not received from Pilot  when deploying shared control plane

### ISSUE THREAD

**Issue by packyzbq** [istio/istio#23487]
**Bug description**
Hi, I'm installing istio between two k8s cluster according to [shared control plane](https://istio.io/docs/setup/install/multicluster/shared/).  Now the istio-ingressgateway on Remote Cluster print error logs:
```
Envoy proxy is NOT ready: config not received from Pilot (is Pilot running?): cds updates: 0 successful, 0 rejected; lds updates: 0 successful, 0 rejected
```
I find the argument `--discoveryAddress` is `istiod-remote.istio-system` and there is a svc named `istiod-remote` in `istio-system`.

Is there anything I need to do?

**Expected behavior**

**Steps to reproduce the bug**

**Version (include the output of `istioctl version --remote` and `kubectl version` and `helm version` if you used Helm)**
- istio version: 1.5.2
- kubectl version
```
Client Version: version.Info{Major:"1", Minor:"18", GitVersion:"v1.18.2", GitCommit:"52c56ce7a8272c798dbc29846288d7cd9fbae032", GitTreeState:"clean", BuildDate:"2020-04-16T11:56:40Z", GoVersion:"go1.13.9", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"17", GitVersion:"v1.17.4", GitCommit:"8d8aa39598534325ad77120c120a22b3a990b5ea", GitTreeState:"clean", BuildDate:"2020-03-12T20:55:23Z", GoVersion:"go1.13.8", Compiler:"gc", Platform:"linux/amd64"}
```

**How was Istio installed?**
I use `istioctl --context=<cluster-context> manifest apply -f <modified-profiles>`
For Main Cluster, the profiles is as below.
```
apiVersion: install.istio.io/v1alpha1
kind: IstioOperator
metadata:
  namespace: istio-system
spec:
  hub: docker.io/istio
  tag: 1.5.2
...
   multiCluster:
        enabled: true
        clusterName: "main-cluster"
      omitSidecarInjectorConfigMap: false
      network: "main-cluster-network"
    meshNetworks:
        main-cluster-network:
          endpoints:
          - fromRegistry: Kubernetes
          gateways:
          - registry_service_name: istio-ingressgateway.istio-system.svc.cluster.local
            port: 443
        remote-cluster-network:
          endpoints:
          - fromRegistry: remote-cluster
          gateways:
          - registry_service_name: istio-ingressgateway.istio-system.svc.cluster.local
            port: 443
 ...
```
And the following is on Remote Cluster
```
apiVersion: install.istio.io/v1alpha1
kind: IstioOperator
metadata:
  namespace: istio-system
spec:
  hub: docker.io/istio
  tag: 1.5.2
...
    multiCluster:
        enabled: true
        clusterName: "remote-cluster"
    network: "remote-cluster-network"
        remotePilotAddress: 10.7.12.185

```
**Environment where bug was observed (cloud vendor, OS, etc)**

**Comment by stevenctl** [comment]
The remotePilotAddress of `10.7.12.185` must be reachable by pods in the remote cluster. Are proper firewall policies in place? Usually the ExternalIP of a LoadBalancer isn't a `10.` address (see [RC1918](https://tools.ietf.org/html/rfc1918)). How did you determine that value for the remotePilotAddress?

Also, if you want you direct pod-to-pod communication without going through a gateway, both clusters should use the same value for `network`.

**Comment by packyzbq** [comment]
> Also, if you want you direct pod-to-pod communication with

Thanks for reply. Actually I have two clusters, the main cluster is 10.7.12.x and the remote cluster is 10.48.51.x. They are not the same network, but they can connect with each other。 The istiod on main cluster is exposed by NodePort service.

**Comment by stevenctl** [comment]
To expose the main cluster's istiod to the remote cluster, you will either need to go through the istio-ingressgateway service, or add an [internal load balancer](https://kubernetes.io/docs/concepts/services-networking/service/#internal-load-balancer) to the istiod service. Then, to get the IP to configure remote cluster, run `kubectl --context=${MAIN_CLUSTER_CTX} -n istio-system get svc $SERVICE -o jsonpath='{.status.loadBalancer.ingress[0].ip}'`, where SERVICE is either istiod, or istio-ingressgateway, depending on your choice of using IngressGateway or ILB. 

The remote envoys will try to connect to istiod-remote.istio-system.svc:15012, where istiod-remote.istio-system.svc will resolve to the IP you set in `remotePilotAddress`. NodePort will give a port between 30000-32767 rather than using the default. Is there a blocker to using ingressgateway or an internal load balancer?

**Comment by GregHanson** [comment]
@packyzbq did you verify that the two cluster have the same certs configured?
```
kubectl --context=${MAIN_CLUSTER_CTX} -n istio-system get configmap istio-ca-root-cert -o yaml
apiVersion: v1
data:
  root-cert.pem: |
    -----BEGIN CERTIFICATE-----
    ABCDEF.....
    ...
    ...
   XYZ
    -----END CERTIFICATE-----
```
Should have the same value in REMOTE cluster:
```
$ kubectl --context=${REMOTE_CLUSTER_CTX} -n istio-system get configmap istio-ca-root-cert -o yaml
apiVersion: v1
data:
  root-cert.pem: |
    -----BEGIN CERTIFICATE-----
    ABCDEF.....
    ...
    ...
   XYZ
    -----END CERTIFICATE-----
```
When you create the cacerts secret you need to make sure the same secret is created in both clusters

**Comment by ImmortanJoe** [comment]
@packyzbq Have you resolve the problem？I have the same question with you. I also use NodePort to expose port for istiod, and I modify the discoveryAddress of the istio-ingressgateway in the remote cluster , but it doesn't work.
```
        - --discoveryAddress
        - istiod-remote.istio-system.svc:31384
```

**Comment by stevenctl** [comment]
Right now, exposing istiod via NodePort won't work for the shared control-plane multicluster model. Is there a reason you must use NodePort over IngressGateway with an external IP?

**Comment by ImmortanJoe** [comment]
Thank you for your reply.
I verified some features on the local virtual machine, there's no loadbakancer.Could you tell me why can't I use NodePort?



---Original---

**Comment by stevenctl** [comment]
> Thank you for your reply. I verified some features on the local virtual machine, there's no loadbakancer.Could you tell me why can't I use NodePort?

You can use NodePort, but you need to set the port to the [discovery address](https://istio.io/docs/reference/config/istio.mesh.v1alpha1/#ProxyConfig). How did you obtain the port `31384` to use in your config? You can get it by running `kubectl --context=primary-cluster-ctx -n istio-system get svc istio-ingressgateway  -o=jsonpath='{.spec.ports[?(@.port==15012)].nodePort}'`, 15012 being the default discovery port.

**Comment by packyzbq** [comment]
@stevenctl 
According to what you say, remote cluster try to connect  main cluster 15012 port by default. How can I change the port so that the remote cluster can connect to the main cluster NodePort ?

**Comment by ImmortanJoe** [comment]
> > Thank you for your reply. I verified some features on the local virtual machine, there's no loadbakancer.Could you tell me why can't I use NodePort?
> 
> You can use NodePort, but you need to set the port to the [discovery address](https://istio.io/docs/reference/config/istio.mesh.v1alpha1/#ProxyConfig). How did you obtain the port `31384` to use in your config? You can get it by running `kubectl --context=primary-cluster-ctx -n istio-system get svc istio-ingressgateway -o=jsonpath='{.spec.ports[?(@.port==15012)].nodePort}'`, 15012 being the default discovery port.

@stevenctl 
[root@node1 ~]# kubectl  -n istio-system get svc istio-ingressgateway -o=jsonpath='{.spec.ports[?(@.port==15012)].nodePort}'
31384

The port is  right, The logs of the primary istiod:
```
2020-05-07T02:37:25.755482Z	info	grpc: Server.Serve failed to complete security handshake from "10.244.2.36:34862": tls: first record does not look like a TLS handshake
2020-05-07T02:37:38.129348Z	info	grpc: Server.Serve failed to complete security handshake from "10.244.2.36:34898": tls: first record does not look like a TLS handshake
2020-05-07T02:38:03.649535Z	info	grpc: Server.Serve failed to complete security handshake from "10.244.2.36:34960": tls: first record does not look like a TLS handshake
```

Details in https://github.com/istio/istio/issues/23595

**Comment by ImmortanJoe** [comment]
> @stevenctl
> According to what you say, remote cluster try to connect main cluster 15012 port by default. How can I change the port so that the remote cluster can connect to the main cluster NodePort ?

@packyzbq 
You could edit the deployment of the remote istio-ingressgateway.

`        
       - --discoveryAddress
        - istiod-remote.istio-system.svc:{your port}
`

**Comment by packyzbq** [comment]
> > @stevenctl
> > According to what you say, remote cluster try to connect main cluster 15012 port by default. How can I change the port so that the remote cluster can connect to the main cluster NodePort ?
> 
> @packyzbq
> You could edit the deployment of the remote istio-ingressgateway.
> 
> ` - --discoveryAddress - istiod-remote.istio-system.svc:{your port}`

I have tried but it didn't work. The logs of istio-ingressgateway is still as follows:
```
Envoy proxy is NOT ready: config not received from Pilot (is Pilot running?): cds updates: 0 successful, 0 rejected; lds updates: 0 successful, 0 rejected
```

**Comment by ImmortanJoe** [comment]
@packyzbq I have solved the problem. 

      - --discoveryAddress - istiod-remote.istio-system.svc:{your port}

{your port} must be 15012.

      # If port is 15012, will use SDS.
      # controlPlaneAuthPolicy is for mounted secrets, will wait for the files.
      controlPlaneAuthPolicy: NONE
      discoveryAddress: istiod-remote.istio-system.svc:15012

I edit the NodePort of the primary cluster to 15012, and the remote istio-ingressgateway worked well.

**Comment by andrew-musoke** [comment]
Hi. I am having the same issue but I am using load balancer instead of nodeport. However, I checked the certs like @GregHanson suggested and they are different. I used the same exact command  and files for the master and remote clusters. Why would the certs become different if cacert is the same? 
`
$ k describe secrets -n istio-system cacerts --context=${REMOTE_CLUSTER_CTX}
Name:         cacerts
Namespace:    istio-system
Labels:       <none>
Annotations:  <none>

Type:  Opaque

Data
====
intermediate.cert.pem:  1541 bytes
intermediate.key.pem:   3243 bytes
root-crt.pem:           1042 bytes
cert-chain.pem:         2583 bytes
`
`
$ k describe secrets -n istio-system cacerts
Name:         cacerts
Namespace:    istio-system
Labels:       <none>
Annotations:  <none>

Type:  Opaque

Data
====
intermediate.cert.pem:  1541 bytes
intermediate.key.pem:   3243 bytes
root-crt.pem:           1042 bytes
cert-chain.pem:         2583 bytes
`

**Comment by andrew-musoke** [comment]
> sadf

**Comment by andrew-musoke** [comment]
Figured it out. The multicluster installation doc missed this portion from saying that the cert files must be named exactly as is.
```
The default Istio CA installation configures the location of certificates and keys based on the predefined secret and file names used in the command below (i.e., secret named cacerts, root certificate in a file named root-cert.pem, Istio CA’s key in ca-key.pem, etc.). You must use these specific secret and file names, or reconfigure Istio’s CA when you deploy Istio.
```

**Comment by 12beseuahmad** [comment]
how this issue got resolved, experiencing the same with istio 1.8, primary remote deployment on different network.

**Comment by zx5435** [comment]
not use auto inject
```
kubectl apply -f a.yaml
```

replace manual to gen new yaml & apply
```
istioctl kube-inject -f a.yaml > b.yaml
kubectl apply -f b.yaml
```

It's solved to me.

I found cacert volume have bug when auto inject, because ServiceAccount not work fine. but manual default use root user.
